pkg/controller/registrysyncer/registrysyncer.go

```diff
@@ -0,0 +1,461 @@
+package registrysyncer
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	imagev1 "github.com/openshift/api/image/v1"
+	"github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	crcontrollerutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	controllerutil "github.com/openshift/ci-tools/pkg/controller/util"
+	"github.com/openshift/ci-tools/pkg/util/imagestreamtagmapper"
+	"github.com/openshift/ci-tools/pkg/util/imagestreamtagwrapper"
+)
+
+const ControllerName = "registry_syncer"
+
+func AddToManager(mgr manager.Manager,
+	apiCIManager manager.Manager,
+	appCIManager manager.Manager,
+	pullSecretGetter func() []byte,
+	excludeImageStreamTags sets.String,
+	excludeImageStreams sets.String,
+	excludeImageStreamNamespaces sets.String,
+	forbiddenRegistries sets.String,
+) error {
+	apiCIControllerName := fmt.Sprintf("%s_api_ci", ControllerName)
+	apiCILog := logrus.WithField("controller", apiCIControllerName)
+	apiCIReconsiler := &reconciler{
+		ctx:                 context.Background(),
+		log:                 apiCILog,
+		registryClient:      imagestreamtagwrapper.MustNew(apiCIManager.GetClient(), apiCIManager.GetCache()),
+		targetRegstryClient: appCIManager.GetClient(),
+		pullSecretGetter:    pullSecretGetter,
+		forbiddenRegistries: forbiddenRegistries,
+	}
+	apiCIController, err := controller.New(apiCIControllerName, mgr, controller.Options{
+		Reconciler: apiCIReconsiler,
+		// We conflict on ImageStream level which means multiple request for imagestreamtags
+		// of the same imagestream will conflict so stay at one worker in order to reduce the
+		// number of errors we see. If we hit performance issues, we will probably need cluster
+		// and/or imagestream level locking.
+		MaxConcurrentReconciles: 1,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %w", err)
+	}
+
+	apiCIObjectFilter, err := testInputImageStreamTagFilterFactory(apiCILog, excludeImageStreamTags, excludeImageStreams, excludeImageStreamNamespaces)
```

Comment:
As mentioned in Slack, I think it would be good to only have a no-op filter here in the beginning and create a follow-up PR to implement the filtering logic.